test/logic/validator_score_v2_logic_test.rb

```diff
@@ -0,0 +1,142 @@
+    p = Pipeline.new(200, @initial_payload)
+                .then(&set_this_batch)
+                .then(&validators_get)
+                .then(&set_validators_groups)
+                .then(&assign_root_distance_score)
+                .then(&save_validators)
+
+    assert_equal 0, p.payload[:validators].find(@v4.id).score_v2.root_distance_score # root dist avg 5
+    assert_equal 0, p.payload[:validators].find(@v5.id).score_v2.root_distance_score # root dist avg 4
+    assert_equal 1, p.payload[:validators].find(@v2.id).score_v2.root_distance_score # root dist avg 3
+    assert_equal 1, p.payload[:validators].find(@v1.id).score_v2.root_distance_score # root dist avg 2
+    assert_equal 2, p.payload[:validators].find(@v3.id).score_v2.root_distance_score # root dist avg 1
+  end
+
+  test "assign_vote_distance_score" do
+    p = Pipeline.new(200, @initial_payload)
+                .then(&set_this_batch)
+                .then(&validators_get)
+                .then(&set_validators_groups)
+                .then(&assign_root_distance_score)
+                .then(&assign_vote_distance_score)
+                .then(&save_validators)
+
+    assert_equal 0, p.payload[:validators].find(@v5.id).score_v2.vote_distance_score # vote dist avg 5
+    assert_equal 0, p.payload[:validators].find(@v4.id).score_v2.vote_distance_score # vote dist avg 4
+    assert_equal 1, p.payload[:validators].find(@v2.id).score_v2.vote_distance_score # vote dist avg 3
+    assert_equal 1, p.payload[:validators].find(@v1.id).score_v2.vote_distance_score # vote dist avg 2
+    assert_equal 2, p.payload[:validators].find(@v3.id).score_v2.vote_distance_score # vote dist avg 1
+  end
+
+  test "assign_skipped_slot_score" do
+    p = Pipeline.new(200, @initial_payload)
+                .then(&set_this_batch)
+                .then(&validators_get)
+                .then(&set_validators_groups)
+                .then(&assign_skipped_slot_score)
+                .then(&save_validators)
+
+    assert_equal 0, p.payload[:validators].find(@v4.id).score_v2.skipped_slot_score # last avg 0.5
+    assert_equal 0, p.payload[:validators].find(@v5.id).score_v2.skipped_slot_score # last avg 0.4
+    assert_equal 1, p.payload[:validators].find(@v2.id).score_v2.skipped_slot_score # last avg 0.3
+    assert_equal 1, p.payload[:validators].find(@v1.id).score_v2.skipped_slot_score # last avg 0.2
+    assert_equal 2, p.payload[:validators].find(@v3.id).score_v2.skipped_slot_score # last avg 0.1
+  end
+
+  test "save_validators" do
+    p = Pipeline.new(200, @initial_payload)
+                .then(&set_this_batch)
+                .then(&validators_get)
+                .then(&set_validators_groups)
+                .then(&assign_root_distance_score)
+                .then(&assign_vote_distance_score)
+                .then(&assign_skipped_slot_score)
+                .then(&save_validators)
+
+    assert_equal 1, p.payload[:validators].find(@v1.id).score_v2.root_distance_score
```

Comment:
You may want to test db values if they were saved properly instead of integer numbers if you save validators in this step, assign scores in the setup to instance variables and reload here (you can also create assertions with values before pipeline run)